### PR TITLE
Add a notion of global environment to Gillian

### PR DESCRIPTION
**Pull request by giltho**
This PR adds a notion of _initialisation data_ to Gillian
When initialised, a memory receives an `init_data` parameter of its desired type, and is abled to initialise itself with context.

For example, in C, this avoids having to create a global environment dynamically. Instead, it is created statically when the file is compiled and then carried around throughout execution.

This looks big, because it filters throughout the entire engine: every time we create a state, we have to give it `init_data`, and every component of the engine has to agree on its type, which creates annoying but necessary type equations.

On the way, I tried to fix an inconsistency or two about the infrastructures, and took a few notes - see #131.

Before this gets merged, I would like to:
- [x] Clean the LAction types to get rid of anything related to Global env core predicates which don't exist anymore
- [x] Add a syntax for init_data in GIL. For now, I'll just add it as a section at the beginning of the file, it will either be whatever text that has to be parsed, or JSON. I'm thinking I'll go with JSON

### CODE REVIEW

**Comment by giltho** [comment]
Ready for review and merge

File: Gillian-C/lib/MonadicSMemory.ml [hunk 1]
```diff
@@ -466,44 +454,56 @@ module Mem = struct
...
+  else Mem.empty
+
+let init genv = { genv; mem = ref (just_functions genv) }
+let clear { genv; _ } = { genv; mem = ref (just_functions genv) }
+
+(* Getters *)
+
+(* Getters *)
+
+(* Getters *)
```
Comment (NatKarmios) [inline Gillian-C/lib/MonadicSMemory.ml:504]: ?

Comment (giltho) [reply Gillian-C/lib/MonadicSMemory.ml:504]: What the fuck hahahaha

**Review by NatKarmios — APPROVED** [review]
I don't see anything immediately horrifying

**Comment by giltho** [comment]
Thank you, removing the weirdness you flagged and merging

**Pull request merged**